[Merged by Bors] - Support multiple signers in beacon

poszu opened the pull request:
## Motivation
Part of https://github.com/spacemeshos/pm/issues/261
Closes #5088

## Changes
- beacon protocol driver supports registering new keys to participate in the protocol
- when a new epoch starts, the protocol selects active signers that will participate
- state is changed to contain the set of active protocol participants. A participant session consists of:
  * signer
  * VRF nonce
- weakcoin supports multiple participants. Each one publishes the same proposal.

## Test Plan
Existing unit-tests were extended to run beacon and weakcoin with multiple signers per node.

poszu committed:
Support multiple signers in beacon

poszu committed:
Satisfy linter

poszu committed:
Fix TestPrune unit test (#5157)

Closes #5156

dependabot[bot] committed:
build(deps): Bump github.com/multiformats/go-multiaddr from 0.11.0 to 0.12.0 (#5159)

Bumps github.com/multiformats/go-multiaddr from 0.11.0 to 0.12.0.

dshulyak commented on beacon/beacon.go:
```diff
@@ -135,22 +120,60 @@ func New(
-			weakcoin.WithLog(pd.logger.WithName("weakCoin")),
-			weakcoin.WithMaxRound(pd.config.RoundsNumber),
-		)
+	if pd.createWeakCoin == nil {
```
it doesn't make sense to have several weak coin instances. there should be on instance that holds state, and routine that goes over all signers and creates proposal for each one

in current implementation i think the simplest would be to split StartRound into two methods, one generic for all signers and another one accepts signer and publishes proposal

poszu replied:
You are right, I refactored the code, but I didn't split the `StartRound` method as it didn't seem necessary. It now takes a slice of participants (vrf signer + vrf nonce).

dshulyak commented on beacon/beacon.go:
```diff
@@ -848,21 +879,42 @@ func (pd *ProtocolDriver) runConsensusPhase(ctx context.Context, epoch types.Epo
+	} else {
+		for _, session := range active {
+			session := session
+			pd.eg.Go(func() error {
```
could you remove pb.eg.Go in sendToGossip?

dshulyak commented on beacon/handlers.go:
```diff
@@ -38,12 +40,21 @@ var (
+			return coin.HandleProposal(ctx, peer, msg)
+		})
+	}
+	pd.mu.Unlock()
```
should RUnlock? but i think this can be refactored to avoid creating multiple weak coin instances

poszu replied:
Good catch, thanks! The sad part is that it means this code is never executed in tests :(

dshulyak commented on beacon/handlers.go:
```diff
@@ -229,17 +240,17 @@ func (pd *ProtocolDriver) HandleFirstVotes(ctx context.Context, peer p2p.Peer, m
-			log.Uint32("current_epoch", uint32(currentEpoch)),
-			log.Uint32("message_epoch", uint32(m.EpochID)))
+			log.FieldNamed("current_epoch", currentEpoch),
+			log.FieldNamed("message_epoch", m.EpochID))
```
FieldNamed is deprecated as it relies on copying whole array and then unpacking original fields

poszu replied:
I don't understand - what array do you have in mind and where is it unpacked? Is the problem that `func (e EpochID) Field()` is called eagerly inside of `FieldNamed()`?

Anyway - if it's deprecated let's mark it so and create an issue to delete it.

dshulyak replied:
i meant this code in spacemesh logger

```go
func (fl FieldLogger) Info(msg string, fields ...LoggableField) {
	fl.l.Info(msg, unpack(append(fields, String("name", fl.name)))...)
}
```

poszu replied:
OK, but this happens anyway, doesn't it?

fasmat replied:
Would it be much effort to migrate `ProtocolDriver` from using `go-spacemesh/log` to using `zap`? I also don't fully understand why loggers are passed around via context, when `ProtocolDriver` already has an instance of a logger?

poszu replied:
Probably not much effort. I will try to migrate this/next week.

dshulyak commented on beacon/handlers.go:
```diff
@@ -229,17 +240,17 @@ func (pd *ProtocolDriver) HandleFirstVotes(ctx context.Context, peer p2p.Peer, m
-			log.Uint32("current_round", uint32(currentRound)),
-			log.Uint32("message_round", uint32(types.FirstRound)))
+			log.FieldNamed("current_round", currentRound),
+			log.FieldNamed("message_round", types.FirstRound))
```
FieldNamed is deprecated

dshulyak committed:
fetch: parametrize success rate threshold (#5161)

this threshold used to determine if peers are seen as equally responsive within certain threshold.
parametrizing it to run some experiments

fasmat commented on beacon/beacon.go:
```diff
@@ -135,22 +120,60 @@ func New(
+	nonce types.VRFPostIndex
+}
+
+func (s *participant) Id() signerID {
+	return signerID(s.signer.NodeID())
+}
+
+type signerID types.NodeID
+
+func (id signerID) Field() log.Field {
+	return log.ShortStringer("id", types.NodeID(id))
+}
```
I'm not sure about introducing a new type here just to change the logging behavour of `NodeID` only for `beacon`. We should consider `(NodeID) Field()` to return a `log.ShortStringer` and have that behaviour everywhere.

poszu replied:
What's wrong with having a private type here? Btw, it changed the key as well (NodeID logs "node_id", not "id").

fasmat replied:
The problem isn't the private type, it's that `beacon` now logs `NodeID` differently from all other components, which makes it harder to read them imo.

poszu replied:
IMHO, it doesn't make sense to log messages per-signer with a "node_id" field, it has nothing to do with "node" anymore (it's not a node ID in this context, there are many IDs governed by a single node).

Perhaps instead, the `NodeID` type should be dropped entirely in favor of a `SignerID` or something similar? But that's a bigger refactor.

fasmat commented on beacon/beacon.go:
```diff
@@ -848,21 +879,42 @@ func (pd *ProtocolDriver) runConsensusPhase(ctx context.Context, epoch types.Epo
+	// First round
+	round := types.FirstRound
+	pd.setRoundInProgress(round)
+	if msg, err := pd.genFirstRoundMsgBody(epoch); err != nil {
+		logger.With().Error("failed to get firstround message", log.Err(err), round)
+	} else {
```
The only reason `pd.genFirstRoundMsgBody` might fail is if the identity isn't active in the given epoch. But `active` already has been checked for that so can this method return an error?

Alternatively `pd.genFirstRoundMsgBody` could just be part of `sendFirstRoundVote` to simplify the logic.

poszu replied:
It's generally a problem with creating too many "small helper methods" that it is hard to follow in what conditions a method is called (for example, what is already asserted). I looked over code again. It turns out `genFirstRoundMsgBody()` is called in 1 place and the `state` for the epoch has already been picked and asserted that it exist. Hence it makes no sense to pick it up again from `pd.states[epoch]` in the first place.

I will refactor the code to pass the once picked state down the stack.

>Alternatively pd.genFirstRoundMsgBody could just be part of sendFirstRoundVote to simplify the logic.

It cannot because:
- this work is common for all signers so it makes no sense to repeat it
- it mutates shared state by sorting incoming proposals
 
That's why I extracted it from `sendFirstRoundVote`.

fasmat replied:
> this work is common for all signers so it makes no sense to repeat it

I initially assumed the method just creates a message and returns -> that doesn't need to be its own method. The problem here is that the method has an additional side effect (sorting proposals) that isn't not obvious from its name or on a first glance when looking at the method.

> it mutates shared state by sorting incoming proposals

That's fair, but then sorting should be done once in a method that's named "sortIncomingProposals" or similar and the instantiation of the Message object can be done in `sendFirstRoundVote`. At least I think that would make it clearer what's happening 🙂

poszu replied:
I removed the method entirely. ~Btw, I discovered that it had a bug - it took a shared Rlock and mutated the data :(~

dshulyak committed:
mesh: refactor Updates handling  (#5160)

still able to reproduce the bug below.  

> 2023-10-02T15:28:14.002+0200 WARN fd68b.sync mesh failed to process layer from sync {"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "sync", "sessionId": "29422935-68d6-47d1-87a8-02293aa181f3", "layer_id": 23104, "errmsg": "requested layer 8063 is before evicted 13102", "name": "sync"}

in the change i removed Results method from tortoise, and instead keeping Updates until mesh applies block and notifies tortoise that the block was applied, only then tortoise is free to evict results from memory. 

OnApplied has layer and opinion in the signature. Opinion is a unique representation of consensus state. If it changed while blocks were applied it means that history changed and we should not evict pending results.

dshulyak committed:
miner: multi signers (#5135)

closes: https://github.com/spacemeshos/go-spacemesh/issues/5087

data is separated into shared data (beacon and active set) and signer specific data. both beacon and activeset are used from shared data, until smeshers generated a reference ballot. once any smesher generated a ballot it will be using data recorded in the reference ballot.

build method now loops over all signers (copied at the start of the layer). there are parts that have to be run once for all signers and parts that makes sense to run in parallel.

serial parts:
- loading share data (beacon and active set)
- deciding on mesh hash
- tally votes & encode votes tortoise calls

parallel parts:
- loading data (it can be also run serially, but it was convenient to run it in parallel)
- computing eligibilities (this is done once per node startup)
- selecting txs
- publishing proposal. this is the most important to avoid blocking serially in Publish while it runs validation

worker pool (errgroup) is limited by number of cores as there is no network requests during parallel work.

fasmat committed:
Specify gcloud version in release and systest jobs (#5162)

## Motivation
Not specifying a version leads the jobs to download the latest version on every build. Since a new version of gcloud is released on average once per week and those installations take up ~550 MB on our runners and are not removed automatically this eventually fills up the disks of the runners.

## Changes
- Specify a gcloud version in `systests` and `release` jobs. This way only that version is downloaded and re-used on every subsequent run.
-  We can manually update gcloud to a newer version if needed.

## Test Plan
n/a

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

fasmat committed:
Try possible fix for flaky TestAdminEvents on Windows (#5165)

## Motivation
Instead of trying to access the local PoET via `0.0.0.0:10010` access it via `127.0.0.1:10010`. The first address seems to cause problems on some Windows systems

## Changes
Updated standalone config to use 127.0.0.1 for local addresses instead of 0.0.0.0

## Test Plan
n/a

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

poszu committed:
Reuse picked epoch state where it makes sense

Avoid infallibile error checks.

poszu committed:
Don't publish proposals on redundant goroutine

poszu committed:
Support multiple signers in weakcoin

poszu committed:
Removed LittleEndian() from vrfSigner interface

poszu committed:
Merge remote-tracking branch 'origin/develop' into 5088-beacon-support-multiple-ids

poszu commented:
bors try

dshulyak commented on beacon/weakcoin/weak_coin.go:
```diff
@@ -193,7 +192,7 @@ func (wc *WeakCoin) FinishEpoch(ctx context.Context, epoch types.EpochID) {
 		logger.With().Fatal("attempted to finish beacon weak coin for the wrong epoch",
 			epoch,
-			log.Stringer("weak_coin_epoch", wc.epoch),
+			log.FieldNamed("weak_coin_epoch", wc.epoch),
```
can you Uint32(), my idea was that we remove FieldNamed over time and then it reduces amount of refactoring

poszu committed:
Fix beacon regression in systests

poszu commented:
bors try

poszu committed:
Remove usages of log.FieldNamed in beacon pkg

poszu commented:
bors merge

poszu commented:
Failure in unrelated UT

bors merge
